text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+- Feature Name: `direct_enum_discriminant`
+- Start Date: 2024-03-16
+- RFC PR: [rust-lang/rfcs#3607](https://github.com/rust-lang/rfcs/pull/3607)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Enable using **`.enum#discriminant`** on values of enum type from safe code in the same module
+to get the numeric value of the variant's discriminant in the numeric type of its `repr`.
+
+# Motivation
+[motivation]: #motivation
+
+Today in Rust you can use `as` casts on *field-less* `enum`s to get their discriminants,
+but as soon as any variant has fields, that's no longer available.
+
+Rust 1.66 stabilized custom discriminants on variants with fields, but as
+[the release post][rust 1.66 blog] said,
+
+> Rust provides no language-level way to access the raw discriminant of an enum with fields.
+> Instead, currently unsafe code must be used to inspect the discriminant of an enum with fields.
+
+[rust 1.66 blog]: https://blog.rust-lang.org/2022/12/15/Rust-1.66.0.html#explicit-discriminants-on-enums-with-fields
+
+As a result, the [documentation for `mem::Discriminant`][discriminant docs] has a section
+about how to write that `unsafe` code, and a bunch of warnings about the different
+*incorrect* ways that must not be used.
+
+[discriminant docs]: https://doc.rust-lang.org/std/mem/fn.discriminant.html#accessing-the-numeric-value-of-the-discriminant
+
+It's technically [possible](https://github.com/rust-lang/rust/pull/106418#issuecomment-1700399884)
+to write a clever enough safe `match` that compiles down to a no-op in order to get at the discriminant,
+but doing so is annoying and fragile.
+
+And accessing the discriminant is quite useful in various places, so it'd be nice for it to be easy.
+
+For example, `#[derive(PartialOrd)]` on an `enum` today uses internal compiler magic to look at discriminants.
+It would be nice for other derives in the ecosystem -- there's a whole bunch of things on `enum`s --
+to be able to look at the discriminants directly too.
```

Comment:
I don't find this *particular* motivation to be particularly strong, as written. Variant discriminants (but not tags) can always be determined syntactically: A variant's discriminant is always one greater than the discriminant of the preceding variant, unless it is explicitly provided, in which case the explicit value is used. If the variant is the first variant and does not have an explicit discriminant, its discriminant is implicitly `0`. Given this, it's reasonably simple for `derive`s to reflect on the discriminants of an enum's variants — no compiler magic needed.